src/can.c

```diff
@@ -548,9 +548,15 @@ static void thingset_can_reqresp_recv_error_callback(int8_t error, struct isotp_
 static void thingset_can_reqresp_sent_callback(int result, void *arg)
 {
     struct thingset_can *ts_can = arg;
-    if (ts_can->request_response.callback != NULL && result != 0) {
-        ts_can->request_response.callback(NULL, 0, 0, result, 0, ts_can->request_response.cb_arg);
+    if (ts_can->request_response.callback != NULL) {
+        ts_can->request_response.callback(NULL, 0, 0, result, ts_can->request_response.can_id,
+                                          ts_can->request_response.cb_arg);
         thingset_can_reset_request_response(&ts_can->request_response);
+        if (result == 0) {
+            /* maintain unlocking semantics of previous iteration of this code */
+            struct shared_buffer *sbuf = thingset_sdk_shared_buffer();
+            k_sem_give(&sbuf->lock);
```

Comment:
I'm wondering why we are unlocking the buffer here at all, as the shared buffer doesn't seem to be used by the `thingset_can_send_inst` function. Only at my phone at the moment, so it's difficult to check the entire code flow...

Reply:
It's somewhat lost in the mists of time now, but I _think_ the rationale was that we only want to unlock the buffer once the message has been successfully sent. I agree that it's a bit odd. In truth, the whole thing probably wants a bit of a rethink, but I figured this approach at least wouldn't make things any worse.